ddtrace/profiling/collector/_traceback.pyx

```diff
@@ -1,3 +1,18 @@
+cpdef _extract_class_name(frame):
+    # type: (...) -> str
+    """Extract class name from a frame, if possible.
+
+    :param frame: The frame object.
+    """
+    try:
+        return frame.f_locals["self"].__class__.__name__
+    except Exception:
+        try:
+            return frame.f_locals["cls"].__name__
+        except Exception:
+            return ""
```

Comment:
Any lookup on a frame object from a class method will be more expensive. Could we avoid raising an exception if we know we are looking at a `classmethod`, i.e. checking the `f_locals` rather than using an exception to use the `cls` value?

Reply:
How about looking at `frame.f_code.co_varnames[0]`? E.g.
~~~ python
if frame.f_code.co_varnames
    argname = frame.f_code.co_varnames[0]
    value = frame.f_locals[argname]
    if argname == "self":
        return object.__getattribute__(type(value), "__name__")  # use type() and object.__getattribute__ to avoid side-effects
    if argname == "cls":
        return object.__getattribute__(value, "__name__")
    return ""
~~~
More generally, one could take `frame.f_code.co_varnames[0]`, when available, and resolve it to the corresponding Python `value` with `f_locals`. Then check if `value` has an attribute named `frame.f_code.co_name` and inspect what kind of attribute it is. If it is a bound method, then `value` is a `self`; if it is a class method, then `value` is a `cls` etc... Not sure it is worth doing all this extra work though, unless it's all done once and the result is cached.